File: src/main/java/com/example/probation/validatiion/validator/EmailValidator.java
package com.example.probation.validatiion.validator;

import com.example.probation.service.UsersService;
import com.example.probation.validatiion.annotation.UniqueEmail;
import lombok.RequiredArgsConstructor;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

@RequiredArgsConstructor
public class EmailValidator implements ConstraintValidator<UniqueEmail, String> {
    private final UsersService usersService;

    @Override
    public void initialize(UniqueEmail uniqueEmail) {
        ConstraintValidator.super.initialize(uniqueEmail);
    }

    @Override
    public boolean isValid(String email, ConstraintValidatorContext context) {
        if (email == null) {
            return false;
        }

        return !usersService.checkEmailExistence(email);
    }
}

File: src/main/kotlin/com/example/probation/validation/validator/EmailValidator.kt
package com.example.probation.validation.validator

import com.example.probation.service.UsersService
import com.example.probation.validation.annotation.UniqueEmail

import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext

class EmailValidator(
    private val usersService: UsersService
) : ConstraintValidator<UniqueEmail, String> {
    override fun isValid(email: String?, context: ConstraintValidatorContext) = if (email == null) {
        false
    } else {
        !usersService.checkEmailExistence(email)
    }

    override fun initialize(uniqueEmail: UniqueEmail) {
        super.initialize(uniqueEmail)
    }
}
